sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala

```diff
@@ -294,7 +296,12 @@ object ShuffleExchangeExec {
           sorter.sort(iter.asInstanceOf[Iterator[UnsafeRow]])
         }
       } else {
-        rdd
+        part match {
+          case partitioner: RangePartitioner[InternalRow @unchecked, _]
+            if partitioner.getSampledArray != null =>
+            sparkContext.parallelize(partitioner.getSampledArray.toSeq, rdd.getNumPartitions)
```

Comment:
```
part match {
          case partitioner: RangePartitioner[InternalRow @unchecked, _]
            if partitioner.getSampledArray != null =>
            sparkContext.parallelize(partitioner.getSampledArray.toSeq, rdd.getNumPartitions)
          case _ => rdd
        }
```
When the optimization works, It will return the parallelized sampled data instead of the rdd. So I keep the number of the partitions same as the rdd's here

Reply:
Then you will change the number of partitions this shuffle requires.

Reply:
every partitionId is in the expected range [0, part.numPartitions - 1]

Reply:
Ah, you are right. The number of partition is not specified here.